plugins/pulp_python/plugins/models.py

```diff
@@ -137,43 +167,60 @@ def checksum(path, algorithm=DEFAULT_CHECKSUM_TYPE):
                 bits = file_handle.read(chunk_size)
         return hasher.hexdigest()
 
-    @staticmethod
-    def _compression_type(path):
+    @property
+    def parsed_version(self):
         """
-        Return the type of compression used in the file at path. Can be '', '.bz2', '.gz', or
-        '.zip'. '' is returned if the file at path matches none of the magic signatures. This
-        algorithm is based on http://stackoverflow.com/a/13044946.
+        Parsed version allows use of < and > operators order versions.
 
-        :param path: The path to the file you wish to test for compression type.
-        :type  path: basestring
-        :return:     File extension used to represent the compression type found at path.
-        :rtype:      basestring
-        """
-        magic_dict = {
-            "\x1f\x8b\x08": ".gz",
-            "\x42\x5a\x68": ".bz2",
-            "\x50\x4b\x03\x04": ".zip"
-        }
-        # We need to read the first four bytes of the file to compare
-        with open(path) as the_file:
-            header = the_file.read(4)
-        for magic, filetype in magic_dict.items():
-            if header.startswith(magic):
-                return filetype
-        return ''
+        :return: version of the package, in comparable form.
+        :rtype:  pkg_resources.SetuptoolsVersion
+        """
+        return pkg_resources.parse_version(self.version)
 
-    @staticmethod
-    def _metadata_label(attribute):
+    @property
+    def src_path(self):
         """
-        Return the label in the PKG-INFO file that corresponds to the given attribute.
+        Returns the relative path to the package bits.
 
-        :param attribute: The attribute on a Package for which you wish to know the PKG-INFO label
-        :type  attribute: basestring
-        :return:          The label in the PKG-INFO file that can be used to get the field
-        :rtype:           basestring
+        :return: relative path to package
+        :rtype:  basestring
+        """
+        return os.path.join('source', self.name[0], self.name, self.filename)
+
+    @property
+    def checksum_url(self):
+        """
+        Adds checksum information to the relative path.
+
+        :return: relative path with checksum information
+        :rtype:  basestring
+        """
+        return '%s#%s=%s' % (self.src_path, self._checksum_type, self._checksum)
+
+    @property
+    def package_specific_metadata(self):
+        """
+        Returns a dictionary containing the subset of metadata that is not duplicated between
+        packages of the same project.
+
+        :return: metadata for package that is not shared with the rest of the project
+        :rtype:  dict
+        """
+        href = '../../../packages/%s' % self.checksum_url
```

Comment:
This looks very "magic-y" to me. The fact that it backs up three directories is a bit of a mystery to me when I read this. My assumption is that this metadata is only valid when published to a certain place. Could you document why it's  three directories and where this metadata is valid?